tools/Makefile

```
@@ -83,44 +43,67 @@ COMPFLAGS = -absname -w +a-4-9-41-42-44-45-48-70 -strict-sequence \
 LINKFLAGS = $(INCLUDES)
 VPATH := $(filter-out -I,$(INCLUDES))
 
+programs_byte := \
+  ocamldep ocamlprof ocamlcp ocamloptp ocamlmklib  \
+  ocamlmktop ocamlcmt dumpobj ocamlobjinfo \
+  primreq stripdebug cmpbyt
+install_files += $(filter $(installed_tools), $(programs_byte))
+programs_opt := $(programs_byte:%=%.opt)
+
 .PHONY: all allopt opt.opt # allopt and opt.opt are synonyms
+all: $(programs_byte)
+opt.opt: $(programs_opt)
 allopt: opt.opt
 
+$(foreach program, $(programs_byte) $(programs_opt),\
+  $(eval $(call PROGRAM_SYNONYM,$(program))))
+
+$(programs_byte:%=%$(EXE)):
+	$(CAMLC) $(LINKFLAGS) -I $(ROOTDIR) -o $@ $(filter-out %.cmi,$^)
+
+$(programs_opt:%=%$(EXE)):
+	$(CAMLOPT_CMD) $(LINKFLAGS) -I $(ROOTDIR) -o $@ $(filter-out %.cmi,$^)
+
+clean::
+	rm -f $(programs_byte) $(programs_byte:%=%.exe)
+	rm -f $(programs_opt) $(programs_opt:%=%.exe)
+
 # The dependency generator
 
-CAMLDEP_OBJ=ocamldep.cmo
-CAMLDEP_IMPORTS= \
+OCAMLDEP = \
   $(ROOTDIR)/compilerlibs/ocamlcommon.cma \
-  $(ROOTDIR)/compilerlibs/ocamlbytecomp.cma
-ocamldep$(EXE): LINKFLAGS += -compat-32
-$(call byte_and_opt,ocamldep,$(CAMLDEP_IMPORTS) $(CAMLDEP_OBJ),)
-ocamldep$(EXE): depend.cmi
-ocamldep.opt$(EXE): depend.cmi
+  $(ROOTDIR)/compilerlibs/ocamlbytecomp.cma \
+  ocamldep.cmo depend.cmi
 
-clean::
-	rm -f ocamldep ocamldep.exe ocamldep.opt ocamldep.opt.exe
+ocamldep$(EXE): LINKFLAGS += -compat-32
+ocamldep$(EXE): $(OCAMLDEP)
+ocamldep.opt$(EXE): $(call byte2native, $(OCAMLDEP))
 
 # The profiler
 
-CSLPROF=ocamlprof.cmo
-CSLPROF_IMPORTS=config.cmo build_path_prefix_map.cmo misc.cmo identifiable.cmo \
+OCAMLPROF=config.cmo build_path_prefix_map.cmo misc.cmo identifiable.cmo \
   numbers.cmo arg_helper.cmo clflags.cmo terminfo.cmo \
   warnings.cmo location.cmo longident.cmo docstrings.cmo \
   syntaxerr.cmo ast_helper.cmo \
   camlinternalMenhirLib.cmo parser.cmo \
   pprintast.cmo \
-  lexer.cmo parse.cmo
+  lexer.cmo parse.cmo ocamlprof.cmo
 
-$(call byte_and_opt,ocamlprof,$(CSLPROF_IMPORTS) profiling.cmo $(CSLPROF),)
+ocamlprof$(EXE): $(OCAMLPROF)
+ocamlprof.opt$(EXE): $(call byte2native, $(OCAMLPROF))
+all: profiling.cmo
+allopt: profiling.cmx
```

```suggestion
opt.opt: profiling.cmx
```
I apologise - the top of the file says that `allopt` is an alias of `opt.opt` and my previous suggestion broke that...